docs/DesignDocs/oe_log_message()_callback_proposal.md

```diff
@@ -0,0 +1,204 @@
+Proposal of Public Callback API for oe_log_message()
+====
+
+# Motivation
+
+This document describes the proposed public callback API for oe_log_message().
+
+There are several problems with the current logging mechanism that prevent end
+users from making full use of logging:
+- Message can't be redirected to different destinations on demand.
+- Lacking of a method for end users to customize messages.
+- It doesn't have a mechanism to communicate between host application and
+oe_log_message() that end users can't pass more information to the message
+function.
+
+The callback API is supposed to solve such problems.
+
+# User Experience
+
+By adding this public API, a user can customize the output of the log message:
+- The output destination can be stdout, stderr or any specified files that the
+destination is adjustable during runtime.
+- The message can be adjusted according to the log level.
+- The user can get at least as much information about the message as internel
+API. For example, log time, enclave side or host side, log level.
+- Finally, the user can define a function to customize the log message and can
+designate the funtion to process all the log messages.
+
+# Specification
+
+Several necessary items should be added to a public header file.
+- Since log level is exposed to end users, the related type should also be exposed:
+```
+typedef enum _oe_log_level
+{
+    OE_LOG_LEVEL_NONE = 0,
+    OE_LOG_LEVEL_FATAL,
+    OE_LOG_LEVEL_ERROR,
+    OE_LOG_LEVEL_WARNING,
+    OE_LOG_LEVEL_INFO,
+    OE_LOG_LEVEL_VERBOSE,
+    OE_LOG_LEVEL_MAX
+} oe_log_level_t;
+```
+- The message function declaration:
+```
+typedef void (*oe_log_callback_t)(
+    void* context,
+    bool is_enclave,
+    const char* time,
+    long int usecs,
```

Comment:
Can this be struct tm* or something similar?   Don't force conversion to string if the app log will just want it in binary form

Reply:
Fixed, now the time is passed in as "struct tm*" and the user can define the time printing format. Thanks!